app/validators/budget_dates_validator.rb

```diff
@@ -0,0 +1,25 @@
+class BudgetDatesValidator < ActiveModel::Validator
+  def validate(record)
+    return unless record.respond_to?(:period_start_date) && record.respond_to?(:period_end_date)
```

Comment:
These guards are diligant 👍 

I do think that the class is explicitly named for the object they should be used for so I think this line isn't obsviously required however I'm not sure of the harm of it. I did notice that there are no tests for either of these guard clauses.

Reply:
@tahb: I found it super hard to write the tests for the two guards on this which makes me question the need for them.

For the case where the object being validated does not respond to `period_start_date` or `period_end_date`: I've created a simple class that has Activerecord validations and uses the `BudgetDatesValidator`. I can then call `be_valid` on it and expect it to still be valid, this works on my assumption that an object like this is valid by default.

For the case where `period_start_date` and `period_end_date` are nil: I still needed a custom object here as a real `Budget` would use additional validation to check for the prescence of the two attributes which means I cannot test the `BudgetDatesValidator` in isolation.

I am not 100% happy with this, I feel like I should add the validation to the Budget model to remove the need for any of the guards.

Would you mind looking it over and letting me know what your thoughts are?

Reply:
I wonder what would happen if we did not try to rescue the misuse of this validator, what would the impact be of failing loudly?

Reply:
@tahb I talked it over with @MariluzHerrera and realised that only a developer could misue the 'wrong object' issue so we got rid of it.

The guard against nil values for `period_start_date` and `period_end_date` is tricky. We are confident that we do not need the guard as we have `prescence_of` validation on the Budget model, but the Shoulda matcher test fails as they do not seem to be able to work along with the `validates_with`. We talked to @lozette as well and decided to keep the guard for nil values even though the test for it seems really complicated and difficult to parse.

You view appreciated!

Reply:
Cripes! I've had a chat with @erbridge as welll and I am still wondering if this just should go in the model as it is so coupled to Budget - I think that is what I want to do to feel good about this…

Reply:
> we decided to keep the guard for nil values even though the test for it seems really complicated and difficult to parse

This is an important point. Having something that is difficult to test like this is a good sign that there's a problem in the design or somewhere in the responsiblities. 

I wonder how much of this problem comes down to the validator trying to do too much. A great place to be is for the validator to do one thing really well. We might achieve this by giving the responsibility of when to call it along with what data over to the Budget itself. We trust the Budget already that we will always be given 2 Date objects for comparison, rather than 2 Strings. If we try and make the validator do more it starts to feel painful.

Does passing that responsibility and removing the guard make you feel any better about this code?

```
validates_with CustomValidator, start: start, end: end, if: { |budget | budget.start.present? && budget.end.present? }
```

To be clear I'm happy for you to move it into budgets or keep it separate if that suggestion works out!

Reply:
@tahb Were you using Jedi mind tricks on me yesterday?! I had exatcly this realisation on the train home. Using conditional validation on the `validates_with` makes this much nicer and feels like the responsibility is (almost) in the right place.

I've added this with a lambda and tested it on the Budget following the reciepe I think shoulda-matchers uses:

- set invalid state
- validate
- check for expected errors

This feels a little odd to me, but I can't think of a better way.

I need to stop thinking about this PR now - it works and is much better than it was, I think it can go in and we can come back as the need for similar validation arises and split it out there?

Reply:
I can confirm I am not a Jedi 😆 Let's merge!